### PR TITLE
SCP-3392 UPLC simplifier

### PR DESCRIPTION
**Pull request by michaelpj**
Don't look here yet, just benchmarking something.

### CODE REVIEW

**Comment by michaelpj** [comment]
The nofib benchmark is rebuilt every time, so it's a test of whether changes to the compiler make things faster. It would be nice to have size and budget outputs for them too, I'll think about adding those.

**Comment by michaelpj** [comment]
Okay, this is looking very good in a lot of cases, including some of the size/budget tests we have. I think we should almost certainly do this. 

TODO:
- Better tests
- Work out what's going on with the queens example in nofib
    - To do that I'm going to do another PR with size/budget tests for the nofib examples
- Try and reduce the duplication somewhat if at all possible :(

**Comment by michaelpj** [comment]
I regenerated the `plutus-use-cases` scripts before and after this, and it's a 6-15% size decrease, which seems good.

**Comment by michaelpj** [comment]
Well that's extremely mysterious. The `nofib` examples _use less budget_ but apparently run for longer :scream: My extremely sketchy guess is that we cut down on the proportion of the time spent doing machine steps, so more of the time is builtins, which have more prediction error? Unclear!

**Comment by michaelpj** [comment]
The mystery deepens: the nofib tests that got faster got (slightly) bigger (which I don't understand at all), and the ones that got slower got smaller! More digging required.

**Comment by michaelpj** [comment]
Looks like the size bump is due to inlining some constructors. In particular, nullary constructors look like `const`, basically, and we do inline that (we did it [here](https://github.com/input-output-hk/plutus/pull/4311), I noted that it was a bit suspicious at the time!). I think we need some arity analysis to make this not suck. For now I'll just turn it off with a note.

**Comment by kwxm** [comment]
> Well that's extremely mysterious. The `nofib` examples _use less budget_ but apparently run for longer scream My extremely sketchy guess is that we cut down on the proportion of the time spent doing machine steps, so more of the time is builtins, which have more prediction error? Unclear!

You can dump the compiled code using `nofib-exe` (see the help) and then run it with `uplc evaluate -t` to get information about how often it sees the different constructors and the different builtins (use the `-1` option to get a count rather than the budget consumed).  That might help.  You might have to do everything using `flat` if the parser can't handle the UPLC source.

**Comment by michaelpj** [comment]
That's more like it.

**Comment by michaelpj** [comment]
Okay, I think this is ready for review now. Seems to be pretty good overall. The main sadness is that I basically had to duplicate the whole inliner, but I tried to keep the divergences to a minimum and label them, so hopefully it won't be *too* bad.

I should probably write a few more tests before we merge, but I think the code is in a good state, and empirically it seems to work pretty well.

File: plutus-tx-plugin/test/Lift/poly.plc.golden [hunk 1]
```diff
@@ -3,16 +3,22 @@
   [
     [
       [
-        (lam
-          Plugin_Data_Spec_Poly1_i0
-          (lam
-            Plugin_Data_Spec_Poly2_i0
+        (force
+          (delay
```
Comment (ak3n) [inline plutus-tx-plugin/test/Lift/poly.plc.golden:7]: Is it okay that `forceDelayCancel` doesn't happen in `test/Lift` tests?

File: plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs [hunk 2]
```diff
@@ -51,16 +52,17 @@ data CompilationOpts = CompilationOpts {
...
     , _coProfile                  :: Bool
-    } deriving (Eq, Show)
+    } deriving (Show)
 
 makeLenses ''CompilationOpts
 
-defaultCompilationOpts :: CompilationOpts
-defaultCompilationOpts = CompilationOpts True False False False 8 True True True False
+defaultCompilationOpts :: CompilationOpts a
+defaultCompilationOpts = CompilationOpts True False False False 12 True True True mempty False
```
Comment (ak3n) [inline plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs:62]: Maybe extract `_soMaxSimplifierIterations` from `defaultSimplifyOpts` here?

**Review by ak3n — APPROVED** [review]

File: plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs:62]: Mmm, but they're different. This is for the PIR compiler, the other is for the UPLC compiler, so (in principle) they shouldn't import each other. They end up being the same, but I think it's easier for now to just have them be quasi-independent. You could *imagine* wanting to tune them separately...

File: plutus-tx-plugin/test/Lift/poly.plc.golden (hunk 1 above)
Comment (michaelpj) [reply plutus-tx-plugin/test/Lift/poly.plc.golden:7]: Yes: at the moment no optimization happens in `Lift`, this just makes that consistent. We probably *should* do some optimization in `Lift`, but that's a separate problem.

File: plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs (hunk 2 above)
Comment (ak3n) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs:62]: Oh, makes sense, thanks.

**Comment by michaelpj** [comment]
Hydra really doesn't like this PR, I'm going to open a new one.

**Pull request closed without merging**